context/secrets.js

```diff
@@ -4,22 +4,33 @@ import React, {
   useMemo,
   useContext,
   useCallback,
+  useEffect,
 } from 'react'
 
-import { upsert, find } from '../lib/secretsManager'
+import SecretsManager from '../lib/secretsManager'
+
+const secretsManager = new SecretsManager()
+
+;(async () => await secretsManager.init())()
```

Comment:
let's avoid this pattern please, find a more idiomatic way to do this instead. if we need really async initialization, put it in application startup and display splash screen until the secrets manager is initialized rather than coming up with fancy ways to do it. plus, this probably doesn't do what you think it does, it most likely suffers from a race condition because the code is not actually waiting for the async initialization to complete before moving on. this probably only seems to work because of timing in how things are executed in the app. make this async initialization take longer, and the app will blow up

Reply:
You're correct, this pattern doesn't make any sense at this place. It is suited though to the useEffect callback body inside the secrets provider function. See: https://github.com/nearform/optic-expo/pull/25#discussion_r580632889 and the following gist:

```js
  // ...
  
  const secretsManager = useMemo(() => new SecretsManager(), [])

  const [secrets, setSecrets] = useState([])

  useEffect(() => {
    ;(async () => {
      await secretsManager.init()
      setSecrets(await secretsManager.find())
    })()
  }, [secretsManager])

 // ...
```

Reply:
aside from the syntax, this approach will still not work. this is not preventing the application from trying to access an uninitialized secrets manager. if the initialization is async and we can't do otherwise, it should be moved to the root of  the component tree, in the entry point of the application, and delay application startup until the initialization is complete, then inject the secretsManager instance via react context. if you put it here, the application may try to operate on an uninitialized instance and there's nothing you can do to prevent that

Reply:
I see what you mean now. I'll fix it right away.